pallets/mips/src/lib.rs

```diff
@@ -1012,61 +1027,79 @@ impl<T: Trait> Module<T> {
+        }
+
+        let voting = Self::voting(id);
+        VoteCount::Success {
+            ayes: voting.ayes_stake,
+            nays: voting.nays_stake,
         }
     }
 
     /// Retrieve proposals made by `address`.
-    pub fn proposed_by(address: T::AccountId) -> Vec<MipsIndex> {
-        Self::proposal_meta()
-            .into_iter()
+    pub fn proposed_by(address: T::AccountId) -> Vec<MipId> {
+        <ProposalMetadata<T>>::iter()
             .filter(|meta| meta.proposer == address)
-            .map(|meta| meta.index)
+            .map(|meta| meta.id)
             .collect()
     }
 
     /// Retrieve proposals `address` voted on
-    pub fn voted_on(address: T::AccountId) -> Vec<MipsIndex> {
-        let mut indices = Vec::new();
-        for meta in Self::proposal_meta().into_iter() {
-            if let Some(votes) = Self::voting(&meta.index) {
-                if votes.ayes.iter().any(|(a, _)| a == &address)
-                    || votes.nays.iter().any(|(a, _)| a == &address)
-                {
-                    indices.push(votes.index);
-                }
-            }
-        }
-        indices
+    pub fn voted_on(address: T::AccountId) -> Vec<MipId> {
+        <ProposalMetadata<T>>::iter()
+            .filter_map(|meta| match Self::proposal_vote(meta.id, &address) {
+                Vote::None => None,
+                _ => Some(meta.id),
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// It generates the next id for proposals and referendums.
+    fn next_mip_id() -> u32 {
+        let id = <MipIdSequence>::get();
+        <MipIdSequence>::put(id + 1);
+
+        id
     }
 
-    /// It generates the next index for proposals and referendums.
-    fn next_index() -> u32 {
-        let index = <SequenceIndex>::get();
-        <SequenceIndex>::put(index + 1);
+    /// It inserts the vote and updates the accountability of target proposal.
+    fn unsafe_vote(id: MipId, proposer: T::AccountId, vote: Vote<BalanceOf<T>>) -> DispatchResult {
+        let mut stats = Self::voting(id);
+
+        match vote {
+            Vote::Yes(deposit) => {
+                stats.ayes_count = stats
+                    .ayes_count
+                    .checked_add(1)
+                    .ok_or_else(|| Error::<T>::NumberOfVotesExceeded)?;
+                stats.ayes_stake = stats
+                    .ayes_stake
+                    .checked_add(&deposit)
+                    .ok_or_else(|| Error::<T>::StakeAmountOfVotesExceeded)?;
+            }
+            Vote::No(deposit) => {
+                stats.nays_count += stats
+                    .nays_count
+                    .checked_add(1)
+                    .ok_or_else(|| Error::<T>::NumberOfVotesExceeded)?;
+                stats.nays_stake += stats
+                    .nays_stake
+                    .checked_add(&deposit)
+                    .ok_or_else(|| Error::<T>::StakeAmountOfVotesExceeded)?;
+            }
+            Vote::None => unreachable!(),
```

Comment:
This can cause panic. In fact, I think we can rename "None" as "Abstain" and make it a feature.

Reply:
It is not possible now because that internal function is called from a public interface which accept a boolean. However, **I agree that a future change could make my initial assumption false**, so I am going to remove the `unreachable!()`, and print a warning. :+1: 

About "Abstain", I love that idea and after a deeper look in the code I found the following:
- We need to change the public interface (it is fast, np).
- I am wondering if anyone is going to pay transaction fees to obtain the same result that they will get if they do nothing.

At this point, @maxsam4  brings an interesting idea: What about non-binary votes? 
For instance,  
``` 
Proposal: update fees at XXX extrinsic
Options:
    a) +5%
    b) +8% 
    c) +10%
```
Do it make sense from business perspective?

Reply:
Probably, but not for V1 ;-)